### PR TITLE
[App] Auto-upgrade / detect environment mis-match from the CLI

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Adds utilities to the lightning app CLI to avoid the need for `python -m XXX`:
- auto-upgrade (after prompt) if a newer version is available
- auto redirect to the current environment (with prompted install if needed) when running lightning out of a different Python from the current one on the path

**Note: The test here is quite minimal for now as this can't be tested without mocking out all of the important pieces (at which point the test wouldn't catch anything)**

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — COMMENTED** [review]
Looks neat so far.

File: src/lightning_app/utilities/cli_helpers.py [hunk 1]
```diff
@@ -214,3 +223,97 @@ def _arrow_time_callback(
...
+        [executable, "-m", "lightning"] + sys.argv[1:],
+        env=os.environ,
+    )
+
+    sys.exit()
+
+
+def _check_version_and_upgrade():
+    new_version = _is_old_version()
+    if new_version:
```
Comment (Borda) [inline src/lightning_app/utilities/cli_helpers.py:270]: does this ask always or only once for each new release?

Comment (ethanwharris) [reply src/lightning_app/utilities/cli_helpers.py:270]: Currently it would ask every time

Comment (ethanwharris) [reply src/lightning_app/utilities/cli_helpers.py:270]: We could maybe cache the response if they choose 'no' and maybe just warn them on future runs

File: src/lightning_app/utilities/cli_helpers.py [hunk 2]
```diff
@@ -214,3 +223,97 @@ def _arrow_time_callback(
...
+                logger.info(f"⚡ RUN: {command}")
+
+                subprocess.run(
+                    [env_executable, "-m"] + command.split(" "),
+                    check=True,
+                )
+            else:
+                sys.exit()
+
+        _redirect_command(env_executable)
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply src/lightning_app/utilities/cli_helpers.py:318]: Discussed offline, let's try this first and can always change it later if users have issues

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/lightning_app/__init__.py [hunk 3]
```diff
@@ -37,6 +37,8 @@
 if module_available("lightning_app.components.demo"):
     from lightning_app.components import demo  # noqa: F401
 
+__package_name__ = "lightning_app".split(".")[0]
```
Comment (Borda) [inline src/lightning_app/__init__.py:40]: ?

File: tests/tests_app/utilities/test_cli_helpers.py [hunk 4]
```diff
@@ -62,3 +63,51 @@ def test_arrow_time_callback():
 
     with pytest.raises(Exception, match="cannot parse time 1 time unit ago"):
         _arrow_time_callback(Mock(), Mock(), "1 time unit ago")
+
+
+@pytest.mark.parametrize(
+    "releases, current_version, newer_version",
```
Comment (Borda) [inline tests/tests_app/utilities/test_cli_helpers.py:69]: how about dev version if it is installed from the source?

**Pull request merged**